Method overloading support for Torchbind

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #51701

**Approach:**

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea.

Differential Revision: [D27701828](https://our.internmc.facebook.com/intern/diff/D27701828)

tugsbayasgalan committed:
[WIP] method overloading

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] method overloading"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] method overloading"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] method overloading"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] method overloading"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] method overloading"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] method overloading"




[ghstack-poisoned]

tugsbayasgalan commented on aten/src/ATen/core/custom_class.cpp:
```diff
@@ -34,21 +34,37 @@ bool isCustomClass(const c10::IValue& v) {
-    return c10::fmap(methods, [](const std::unique_ptr<jit::Function>& fn) {
-      return fn->getSchema();
-    });
+    auto& custom_class_methods = customClassMethods();
```
Not sure what to do with this?

tugsbayasgalan commented on aten/src/ATen/core/type.cpp:
```diff
@@ -1385,11 +1409,39 @@ torch::jit::Function& ClassType::getMethod(const std::string& name) const {
+  return nullptr;
+}
+
+// std::vector<torch::jit::Function*> ClassType::getOverloadedMethod(const std::string& name) const {
```
similar to getMethod, but we don't need it. It is because getMethod returns function reference and it is not possible to return the vector of function reference in C++

tugsbayasgalan committed:
Update on "[WIP] method overloading"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] method overloading"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] method overloading"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] method overloading"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] method overloading"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] method overloading"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] method overloading"




[ghstack-poisoned]

eellison commented on torch/csrc/jit/frontend/sugared_value.h:
```diff
@@ -458,15 +458,37 @@ struct MethodValue : public SugaredValue {
+          // When we import the graph back (serialization), we will sometimes
+          // receive mangled names. Therefore we need to make sure we don't add
+          // extra __[number] tag again.
+          auto mangled_name = method_name;
```
I think this is a sign of something we need to clean up. When we serialize an invoked overloaded method, we should be serializing the qualified name of a particular overload. That particular overload should be an actual method on the class which gets serialized correctly, and the OverloadedFunction should have a reference to the list of methods which it resolves to.

gmagogsfm replied:
+1

tugsbayasgalan replied:
I changed it so that we only serialize only the qualified name. Could you guys take a look? It makes the serialization code bit more complex but at least we don't have to deal with redundant mangling anymore

eellison commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -2247,6 +2256,14 @@ struct TORCH_API ClassType : public NamedType {
+      overloaded_methods_;
+
+  // Map of mangled name to function pointer
+  std::unordered_map<std::string, torch::jit::Function*> mangled_to_function_;
```
i think the mangled names should be stored normally as all the other functions

tugsbayasgalan replied:
Yeah this is just to keep track of what names maps to what instead of having to go through entire list of methods

eellison reviewed (commented):
looks good!! I didn't do a thorough review, just commenting on one particular thing that stood out

gmagogsfm commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -1038,6 +1039,26 @@ inline py::object invokeScriptMethodFromPython(
+          kwargs,
+          method.owner()._ivalue());
+      return method;
+    } catch (...) {
```
I think we should only catch `schema_match_error` here rather than any generic exception

tugsbayasgalan replied:
done

gmagogsfm reviewed (commented):
As a reminder, I think we should write some documentations on TorchBind after this batch of new features land.

tugsbayasgalan commented on aten/src/ATen/core/overloaded_function.h:
```diff
@@ -0,0 +1,175 @@
+  c10::intrusive_ptr<c10::ivalue::Future> runAsync(
+      Stack& stack,
+      TaskLauncher /* not used */) override {
+    run(stack);
```
This was copied from BuiltinOpFunction, we never use this so we can delete it

tugsbayasgalan replied:
Seems like it is a required method for subclasses for Function struct.

SplitInfinity reviewed (commented):
Found some old cached review I forgot to submit, some of the comments might be out of date.

tugsbayasgalan committed:
Update on "[WIP] method overloading"


**Approach:**

These are the main changes in this PR: 

**Loading Schemas inside OverloadedFunction:**

We only can query possible schemas for overloaded function inside overloaded_function.h (aka can't do schema matching inside overloaded_function.h) for following two reasons:

1. For schema matching with python arguments, we still have to define the method signature in the header file which doesn't allow python dependencies. 

2. For schema matching with Torchscript arguments, we can't use methods defined in FunctionSchema due to circular import issues. 

Therefore, we can only query possible schemas inside overloaded_function and pass it to schema matcher to do the rest of the work. 

**Mangled Names**

Interpreter looks up methods by their names through a Node's attribute. Hence, even if we do schema matching cleanly inside sugarad_value.h or overloaded_function.h, we still run into issue of figuring out what method to call in the interpreter. The work around for this is to mangle overloaded names so that we can uniquely determine which overloaded method to call. (cc @eellison)  (One work around for this might be modifying get_method to take an optional parameter that tells which overloaded function it is using and index into the overloaded_methods vector. This will require us to define another attribute inside Node to carry that extra number). 

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea. 


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] method overloading"


**Approach:**

These are the main changes in this PR: 

**Loading Schemas inside OverloadedFunction:**

We only can query possible schemas for overloaded function inside overloaded_function.h (aka can't do schema matching inside overloaded_function.h) for following two reasons:

1. For schema matching with python arguments, we still have to define the method signature in the header file which doesn't allow python dependencies. 

2. For schema matching with Torchscript arguments, we can't use methods defined in FunctionSchema due to circular import issues. 

Therefore, we can only query possible schemas inside overloaded_function and pass it to schema matcher to do the rest of the work. 

**Mangled Names**

Interpreter looks up methods by their names through a Node's attribute. Hence, even if we do schema matching cleanly inside sugarad_value.h or overloaded_function.h, we still run into issue of figuring out what method to call in the interpreter. The work around for this is to mangle overloaded names so that we can uniquely determine which overloaded method to call. (cc @eellison)  (One work around for this might be modifying get_method to take an optional parameter that tells which overloaded function it is using and index into the overloaded_methods vector. This will require us to define another attribute inside Node to carry that extra number). 

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea. 


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] method overloading"


**Approach:**

These are the main changes in this PR: 

**Loading Schemas inside OverloadedFunction:**

We only can query possible schemas for overloaded function inside overloaded_function.h (aka can't do schema matching inside overloaded_function.h) for following two reasons:

1. For schema matching with python arguments, we still have to define the method signature in the header file which doesn't allow python dependencies. 

2. For schema matching with Torchscript arguments, we can't use methods defined in FunctionSchema due to circular import issues. 

Therefore, we can only query possible schemas inside overloaded_function and pass it to schema matcher to do the rest of the work. 

**Mangled Names**

Interpreter looks up methods by their names through a Node's attribute. Hence, even if we do schema matching cleanly inside sugarad_value.h or overloaded_function.h, we still run into issue of figuring out what method to call in the interpreter. The work around for this is to mangle overloaded names so that we can uniquely determine which overloaded method to call. (cc @eellison)  (One work around for this might be modifying get_method to take an optional parameter that tells which overloaded function it is using and index into the overloaded_methods vector. This will require us to define another attribute inside Node to carry that extra number). 

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea. 


[ghstack-poisoned]

tugsbayasgalan commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -56,6 +57,11 @@
 namespace torch {
 namespace jit {
 
+struct python_args {
```
@SplitInfinity is this what you had in mind?

tugsbayasgalan committed:
Update on "[WIP] method overloading"


**Approach:**

These are the main changes in this PR: 

**Loading Schemas inside OverloadedFunction:**

We only can query possible schemas for overloaded function inside overloaded_function.h (aka can't do schema matching inside overloaded_function.h) for following two reasons:

1. For schema matching with python arguments, we still have to define the method signature in the header file which doesn't allow python dependencies.  (EDIT: I think it might be solved now. Let's see if it passes the tests. I created a struct that wraps python args inside pybind_utils and forward declare it in method.h)

2. For schema matching with Torchscript arguments, we can't use methods defined in FunctionSchema due to circular import issues. (EDIT: I think it might be solved now. Let's see if it passes the tests)


**Mangled Names**

Interpreter looks up methods by their names through a Node's attribute. Hence, even if we do schema matching cleanly inside sugarad_value.h or overloaded_function.h, we still run into issue of figuring out what method to call in the interpreter. The work around for this is to mangle overloaded names so that we can uniquely determine which overloaded method to call. (cc @eellison)  (One work around for this might be modifying get_method to take an optional parameter that tells which overloaded function it is using and index into the overloaded_methods vector. This will require us to define another attribute inside Node to carry that extra number). 

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea. 


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] method overloading"


**Approach:**

These are the main changes in this PR: 

**Loading Schemas inside OverloadedFunction:**

We only can query possible schemas for overloaded function inside overloaded_function.h (aka can't do schema matching inside overloaded_function.h) for following two reasons:

1. For schema matching with python arguments, we still have to define the method signature in the header file which doesn't allow python dependencies.  (EDIT: I think it might be solved now. Let's see if it passes the tests. I created a struct that wraps python args inside pybind_utils and forward declare it in method.h)

2. For schema matching with Torchscript arguments, we can't use methods defined in FunctionSchema due to circular import issues. (EDIT: I think it might be solved now. Let's see if it passes the tests)


**Mangled Names**

Interpreter looks up methods by their names through a Node's attribute. Hence, even if we do schema matching cleanly inside sugarad_value.h or overloaded_function.h, we still run into issue of figuring out what method to call in the interpreter. The work around for this is to mangle overloaded names so that we can uniquely determine which overloaded method to call. (cc @eellison)  (One work around for this might be modifying get_method to take an optional parameter that tells which overloaded function it is using and index into the overloaded_methods vector. This will require us to define another attribute inside Node to carry that extra number). 

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea. 


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] method overloading"


**Approach:**

These are the main changes in this PR: 

**Loading Schemas inside OverloadedFunction:**

We only can query possible schemas for overloaded function inside overloaded_function.h (aka can't do schema matching inside overloaded_function.h) for following two reasons:

1. For schema matching with python arguments, we still have to define the method signature in the header file which doesn't allow python dependencies.  (EDIT: I think it might be solved now. Let's see if it passes the tests. I created a struct that wraps python args inside pybind_utils and forward declare it in method.h)

2. For schema matching with Torchscript arguments, we can't use methods defined in FunctionSchema due to circular import issues. (EDIT: I think it might be solved now. Let's see if it passes the tests)


**Mangled Names**

Interpreter looks up methods by their names through a Node's attribute. Hence, even if we do schema matching cleanly inside sugarad_value.h or overloaded_function.h, we still run into issue of figuring out what method to call in the interpreter. The work around for this is to mangle overloaded names so that we can uniquely determine which overloaded method to call. (cc @eellison)  (One work around for this might be modifying get_method to take an optional parameter that tells which overloaded function it is using and index into the overloaded_methods vector. This will require us to define another attribute inside Node to carry that extra number). 

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea. 


[ghstack-poisoned]

gmagogsfm commented on aten/src/ATen/core/custom_class.cpp:
```diff
@@ -34,21 +34,31 @@ bool isCustomClass(const c10::IValue& v) {
-  customClassMethods().emplace_back(std::move(fn));
+  auto& custom_class_methods = customClassMethods();
+  auto it =
+      custom_class_methods.insert(std::pair<std::string, method_overloads_list>(
```
What if the key already exists in map?

tugsbayasgalan replied:
Now it checks beforehand.

gmagogsfm commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -1075,10 +1075,10 @@ struct PythonPrintImpl {
-          TORCH_INTERNAL_ASSERT(
-              method.qualname() ==
-              QualifiedName(selfClass->name()->qualifiedName(), methodName));
+          // const Function& method = selfClass->getMethod(node->s(attr::name));
+          // TORCH_INTERNAL_ASSERT(
+          //     method.qualname() ==
+          //     QualifiedName(selfClass->name()->qualifiedName(), methodName));
```
Is this intended?

tugsbayasgalan replied:
This is changed to reflect how we serialize overloaded method

gmagogsfm commented on aten/src/ATen/core/type.cpp:
```diff
@@ -1071,6 +1071,17 @@ void ClassType::addMethod(torch::jit::Function* method) {
 }
 
+void ClassType::addOverloadedMethod(torch::jit::Function* method) {
+  auto it = overloaded_methods_.insert(
```
similar here, guard against the case where `method_name` already exists as key in map.

tugsbayasgalan replied:
yep.

tugsbayasgalan replied:
It checks now.

gmagogsfm commented on test/jit/test_torchbind.py:
```diff
@@ -327,6 +327,22 @@ def forward(self):
+
+    def test_torchbind_overloaded_method_init(self):
+        def f():
+            val = torch.classes._TorchScriptTesting._FooOverloadInit(3, 4, "hi")
```
Can you test overloaded init that has slightly different behavior, so that we can be sure the right method is called? Also I would recommend calling both init variants here to make sure they both work.

tugsbayasgalan committed:
Update on "[WIP] method overloading"


**Approach:**

These are the main changes in this PR: 

**Loading Schemas inside OverloadedFunction:**

We only can query possible schemas for overloaded function inside overloaded_function.h (aka can't do schema matching inside overloaded_function.h) for following two reasons:

1. For schema matching with python arguments, we still have to define the method signature in the header file which doesn't allow python dependencies.  (EDIT: I think it might be solved now. Let's see if it passes the tests. I created a struct that wraps python args inside pybind_utils and forward declare it in method.h)

2. For schema matching with Torchscript arguments, we can't use methods defined in FunctionSchema due to circular import issues. (EDIT: I think it might be solved now. Let's see if it passes the tests)


**Mangled Names**

Interpreter looks up methods by their names through a Node's attribute. Hence, even if we do schema matching cleanly inside sugarad_value.h or overloaded_function.h, we still run into issue of figuring out what method to call in the interpreter. The work around for this is to mangle overloaded names so that we can uniquely determine which overloaded method to call. (cc @eellison)  (One work around for this might be modifying get_method to take an optional parameter that tells which overloaded function it is using and index into the overloaded_methods vector. This will require us to define another attribute inside Node to carry that extra number). 

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea. 


[ghstack-poisoned]

tugsbayasgalan commented on aten/src/ATen/core/type.cpp:
```diff
@@ -1070,6 +1070,19 @@ void ClassType::addMethod(torch::jit::Function* method) {
+  const std::string& mangled_name = method->name() + "__" +
+      std::to_string(overloaded_methods_[method->name()].size());
+
+  overloaded_methods_[method->name()].push_back(method);
```
done

tugsbayasgalan commented on aten/src/ATen/core/custom_class.cpp:
```diff
@@ -34,21 +34,34 @@ bool isCustomClass(const c10::IValue& v) {
-  customClassMethods().emplace_back(std::move(fn));
+  auto& custom_class_methods = customClassMethods();
+  if (custom_class_methods.find(fn->name()) == custom_class_methods.end()) {
+    custom_class_methods[fn->name()] = method_list();
```
done

tugsbayasgalan commented on aten/src/ATen/core/custom_class.cpp:
```diff
@@ -34,21 +34,34 @@ bool isCustomClass(const c10::IValue& v) {
 
-std::vector<std::unique_ptr<jit::Function>>& customClassMethods() {
-  static std::vector<std::unique_ptr<jit::Function>> customClassMethods;
+using method_list = std::vector<std::unique_ptr<jit::Function>>;
```
done

tugsbayasgalan committed:
Update on "[WIP] method overloading"


**Approach:**

These are the main changes in this PR: 

**Loading Schemas inside OverloadedFunction:**

We only can query possible schemas for overloaded function inside overloaded_function.h (aka can't do schema matching inside overloaded_function.h) for following two reasons:

1. For schema matching with python arguments, we still have to define the method signature in the header file which doesn't allow python dependencies.  (EDIT: I think it might be solved now. Let's see if it passes the tests. I created a struct that wraps python args inside pybind_utils and forward declare it in method.h)

2. For schema matching with Torchscript arguments, we can't use methods defined in FunctionSchema due to circular import issues. (EDIT: I think it might be solved now. Let's see if it passes the tests)


**Mangled Names**

Interpreter looks up methods by their names through a Node's attribute. Hence, even if we do schema matching cleanly inside sugarad_value.h or overloaded_function.h, we still run into issue of figuring out what method to call in the interpreter. The work around for this is to mangle overloaded names so that we can uniquely determine which overloaded method to call. (cc @eellison)  (One work around for this might be modifying get_method to take an optional parameter that tells which overloaded function it is using and index into the overloaded_methods vector. This will require us to define another attribute inside Node to carry that extra number). 

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea.

Differential Revision: [D27701828](https://our.internmc.facebook.com/intern/diff/D27701828)

[ghstack-poisoned]

gmagogsfm commented on test/jit/test_torchbind.py:
```diff
@@ -111,7 +111,7 @@ def foo():
         # rather than calling the __init__wrapper nonsense
         fc = FileCheck().check('prim::CreateObject()')\
-                        .check('prim::CallMethod[name="__init__"]')
+                        .check('prim::CallMethod[name="__init____0"]')
```
Will this be fragile? Maybe we can just check for 
```
.check('prim::CallMethod[name="__init__')
```

tugsbayasgalan replied:
I think the name of call method will still be __init____0 tho. Are you saying we should just check if this method name has a substring "__init__" ?

tugsbayasgalan replied:
Apparently there is no method on FileCheck that does substring check.

gmagogsfm replied:
FileCheck is regex-based. If you don't include the closing quotation mark and bracket, it should work.

tugsbayasgalan replied:
oh lol my bad

tugsbayasgalan replied:
done

gmagogsfm commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -1066,26 +1067,33 @@ struct PythonPrintImpl {
+          if (auto overloadedMethod =
+                  dynamic_cast<OverloadedFunction*>(method)) {
+            stmt << "(" << useOf(self) << ")"
+                 << "." << method->qualname().name() << "(";
```
Is this unique for each overloaded variant?

tugsbayasgalan replied:
This will go through sugared_value.h again to do schema matching. Therefore, it doesn't matter if the name is not unique i think

tugsbayasgalan commented on aten/src/ATen/core/custom_class.cpp:
```diff
@@ -34,21 +34,49 @@ bool isCustomClass(const c10::IValue& v) {
+    for (auto& method_it : methods.second) {
+      method_vector.push_back(std::move(method_it));
+    }
+  }
```
Ohh i think u looked at the older version of this PR. This has been removed.

tugsbayasgalan committed:
Update on "[WIP] method overloading"


**Approach:**

These are the main changes in this PR: 

**Loading Schemas inside OverloadedFunction:**

We only can query possible schemas for overloaded function inside overloaded_function.h (aka can't do schema matching inside overloaded_function.h) for following two reasons:

1. For schema matching with python arguments, we still have to define the method signature in the header file which doesn't allow python dependencies.  (EDIT: I think it might be solved now. Let's see if it passes the tests. I created a struct that wraps python args inside pybind_utils and forward declare it in method.h)

2. For schema matching with Torchscript arguments, we can't use methods defined in FunctionSchema due to circular import issues. (EDIT: I think it might be solved now. Let's see if it passes the tests)


**Mangled Names**

Interpreter looks up methods by their names through a Node's attribute. Hence, even if we do schema matching cleanly inside sugarad_value.h or overloaded_function.h, we still run into issue of figuring out what method to call in the interpreter. The work around for this is to mangle overloaded names so that we can uniquely determine which overloaded method to call. (cc @eellison)  (One work around for this might be modifying get_method to take an optional parameter that tells which overloaded function it is using and index into the overloaded_methods vector. This will require us to define another attribute inside Node to carry that extra number). 

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea.

Differential Revision: [D27701828](https://our.internmc.facebook.com/intern/diff/D27701828)

[ghstack-poisoned]

tugsbayasgalan commented on aten/src/ATen/core/custom_class.cpp:
```diff
@@ -34,21 +34,49 @@ bool isCustomClass(const c10::IValue& v) {
 
-std::vector<std::unique_ptr<jit::Function>>& customClassMethods() {
-  static std::vector<std::unique_ptr<jit::Function>> customClassMethods;
+using method_overloads_list = std::vector<std::unique_ptr<jit::Function>>;
```
done

tugsbayasgalan committed:
Update on "Method overloading support for Torchbind "


**Approach:**

These are the main changes in this PR: 

**Loading Schemas inside OverloadedFunction:**

We only can query possible schemas for overloaded function inside overloaded_function.h (aka can't do schema matching inside overloaded_function.h) for following two reasons:

1. For schema matching with python arguments, we still have to define the method signature in the header file which doesn't allow python dependencies.  (EDIT: I think it might be solved now. Let's see if it passes the tests. I created a struct that wraps python args inside pybind_utils and forward declare it in method.h)

2. For schema matching with Torchscript arguments, we can't use methods defined in FunctionSchema due to circular import issues. (EDIT: I think it might be solved now. Let's see if it passes the tests)


**Mangled Names**

Interpreter looks up methods by their names through a Node's attribute. Hence, even if we do schema matching cleanly inside sugarad_value.h or overloaded_function.h, we still run into issue of figuring out what method to call in the interpreter. The work around for this is to mangle overloaded names so that we can uniquely determine which overloaded method to call. (cc @eellison)  (One work around for this might be modifying get_method to take an optional parameter that tells which overloaded function it is using and index into the overloaded_methods vector. This will require us to define another attribute inside Node to carry that extra number). 

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea.

Differential Revision: [D27701828](https://our.internmc.facebook.com/intern/diff/D27701828)

[ghstack-poisoned]

tugsbayasgalan commented on torch/csrc/jit/api/module.cpp:
```diff
@@ -2,6 +2,7 @@
 
 #include <ATen/record_function.h>
 #include <c10/util/Exception.h>
+#include <c10/util/Optional.h>
```
ohh will be removed

tugsbayasgalan commented on torch/csrc/jit/frontend/sugared_value.h:
```diff
@@ -458,15 +458,27 @@ struct MethodValue : public SugaredValue {
-          throw ErrorReport(loc)
-              << " method '" << method.name() << "' is called recursively. "
-              << "Recursive calls are not supported";
+        Function* method = class_type->findMethod(method_name);
+        if (auto overloaded_method =
+                dynamic_cast<OverloadedFunction*>(method)) {
```
We register the mangled names when we call addOverloadedMethod. I will update with a comment inside addOverloadedMethod.

tugsbayasgalan commented on torch/csrc/jit/frontend/sugared_value.h:
```diff
@@ -458,15 +458,27 @@ struct MethodValue : public SugaredValue {
+        Function* method = class_type->findMethod(method_name);
+        if (auto overloaded_method =
+                dynamic_cast<OverloadedFunction*>(method)) {
+          auto matched = overloaded_method->matchOverloadedSchemas(
+              loc, class_type, *f.graph(), argsWithSelf, kwargs);
```
I don't remember either but I can give it another try.

tugsbayasgalan replied:
I think this is because OverloadedFunction has to implement virtual methods defined in Function class. In the case of OverloadedFunction storing other functions inside, it is hard to define the behaviour of methods like run(). We can maybe make OverloadedFunction not subclass from base Function and just keep it as a seperate struct.

tugsbayasgalan replied:
Ok I added classType as a field inside OverloadedFunction so it knows the other overloaded functions through its' owning class. I think this is similar to how Method object is implemented right now.

tugsbayasgalan committed:
Update on "Method overloading support for Torchbind "


**Approach:**

These are the main changes in this PR: 

**Loading Schemas inside OverloadedFunction:**

We only can query possible schemas for overloaded function inside overloaded_function.h (aka can't do schema matching inside overloaded_function.h) for following two reasons:

1. For schema matching with python arguments, we still have to define the method signature in the header file which doesn't allow python dependencies.  (EDIT: I think it might be solved now. Let's see if it passes the tests. I created a struct that wraps python args inside pybind_utils and forward declare it in method.h)

2. For schema matching with Torchscript arguments, we can't use methods defined in FunctionSchema due to circular import issues. (EDIT: I think it might be solved now. Let's see if it passes the tests)


**Mangled Names**

Interpreter looks up methods by their names through a Node's attribute. Hence, even if we do schema matching cleanly inside sugarad_value.h or overloaded_function.h, we still run into issue of figuring out what method to call in the interpreter. The work around for this is to mangle overloaded names so that we can uniquely determine which overloaded method to call. (cc @eellison)  (One work around for this might be modifying get_method to take an optional parameter that tells which overloaded function it is using and index into the overloaded_methods vector. This will require us to define another attribute inside Node to carry that extra number). 

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea.

Differential Revision: [D27701828](https://our.internmc.facebook.com/intern/diff/D27701828)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Method overloading support for Torchbind "


**Approach:**

These are the main changes in this PR: 

**Schema matching inside OverloadedFunction:**

By registering custom class type (aka ClassType) as a field inside OverloadedFunction, we can query other overloaded functions with the same name. Using this, we can do schema matching internally inside OverloadedFunction. 

**Mangled Names**

Interpreter looks up methods by their names through a Node's attribute. Hence, even if we do schema matching cleanly inside sugarad_value.h or overloaded_function.h, we still run into issue of figuring out what method to call in the interpreter. The work around for this is to mangle overloaded names so that we can uniquely determine which overloaded method to call. (cc @eellison) 

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea.

Differential Revision: [D27701828](https://our.internmc.facebook.com/intern/diff/D27701828)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Method overloading support for Torchbind "


**Approach:**

These are the main changes in this PR: 

**Schema matching inside OverloadedFunction:**

By registering custom class type (aka ClassType) as a field inside OverloadedFunction, we can query other overloaded functions with the same name. Using this, we can do schema matching internally inside OverloadedFunction. 

**Mangled Names**

Interpreter looks up methods by their names through a Node's attribute. Hence, even if we do schema matching cleanly inside sugarad_value.h or overloaded_function.h, we still run into issue of figuring out what method to call in the interpreter. The work around for this is to mangle overloaded names so that we can uniquely determine which overloaded method to call. (cc @eellison) 

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea.

Differential Revision: [D27701828](https://our.internmc.facebook.com/intern/diff/D27701828)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Method overloading support for Torchbind "


**Approach:**

These are the main changes in this PR: 

**Schema matching inside OverloadedFunction:**

By registering custom class type (aka ClassType) as a field inside OverloadedFunction, we can query other overloaded functions with the same name. Using this, we can do schema matching internally inside OverloadedFunction. 

**Mangled Names**

Interpreter looks up methods by their names through a Node's attribute. Hence, even if we do schema matching cleanly inside sugarad_value.h or overloaded_function.h, we still run into issue of figuring out what method to call in the interpreter. The work around for this is to mangle overloaded names so that we can uniquely determine which overloaded method to call. (cc @eellison) 

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea.

Differential Revision: [D27701828](https://our.internmc.facebook.com/intern/diff/D27701828)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27701828).

tugsbayasgalan committed:
Update on "Method overloading support for Torchbind "


**Approach:**

These are the main changes in this PR: 

**Schema matching inside OverloadedFunction:**

By registering custom class type (aka ClassType) as a field inside OverloadedFunction, we can query other overloaded functions with the same name. Using this, we can do schema matching internally inside OverloadedFunction. 

**Mangled Names**

Interpreter looks up methods by their names through a Node's attribute. Hence, even if we do schema matching cleanly inside sugarad_value.h or overloaded_function.h, we still run into issue of figuring out what method to call in the interpreter. The work around for this is to mangle overloaded names so that we can uniquely determine which overloaded method to call. (cc @eellison) 

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea.

Differential Revision: [D27701828](https://our.internmc.facebook.com/intern/diff/D27701828)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27701828).

tugsbayasgalan committed:
Update on "Method overloading support for Torchbind "


**Approach:**

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea.

Differential Revision: [D27701828](https://our.internmc.facebook.com/intern/diff/D27701828)

[ghstack-poisoned]

SplitInfinity commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -2125,6 +2125,22 @@ struct TORCH_API ClassType : public NamedType {
+
+  // given the mangled name, return the function pointer
+  // corresponding that name.
+  torch::jit::Function* getMangledOverloadedMethod(
+      const std::string& name) const;
```
Same here, can this be private?

tugsbayasgalan replied:
Done

tugsbayasgalan committed:
Update on "Method overloading support for Torchbind "


**Approach:**

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea.

Differential Revision: [D27701828](https://our.internmc.facebook.com/intern/diff/D27701828)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27701828).

tugsbayasgalan commented on aten/src/ATen/core/type.cpp:
```diff
@@ -1408,6 +1436,23 @@ torch::jit::Function& ClassType::getMethod(const std::string& name) const {
+  return c10::nullopt;
+}
+
+torch::jit::Function* ClassType::getMangledOverloadedMethod(
+    const std::string& name) const {
+  if (mangled_to_function_.find(name) != mangled_to_function_.end()) {
+    auto actual_method_idx = mangled_to_function_.find(name)->second;
+    return methods_[actual_method_idx];
+  }
+  return nullptr;
+}
```
Done

tugsbayasgalan commented on torch/custom_class.h:
```diff
@@ -456,14 +455,17 @@ class class_ {
     // ClassTypes do not hold ownership of their methods (normally it
     // those are held by the CompilationUnit), so we need a proxy for
     // that behavior here.
     auto method_val = method.get();
-    classTypePtr->addMethod(method_val);
+    classTypePtr->addOverloadedMethod(method.get());
```
I think we might have to do one pass to cleanup. But it is hard to know when the user stopped registering methods. Maybe we can address it in a follow up PR?

SplitInfinity replied:
Hmm, okay, makes sense

tugsbayasgalan commented on aten/src/ATen/core/type.cpp:
```diff
@@ -1086,6 +1087,28 @@ void ClassType::addMethod(torch::jit::Function* method) {
+      std::pair<std::string, std::vector<std::string>>(
+          method->name(), std::vector<std::string>()));
+  // create a mangled name for this function and bookkeep
+  // the mangled name and its' corresponding function.
```
done

tugsbayasgalan commented on torch/csrc/jit/frontend/sugared_value.cpp:
```diff
@@ -152,6 +153,10 @@ std::shared_ptr<SugaredValue> SimpleValue::attr(
   } else if (auto classType = value_->type()->cast<ClassType>()) {
     // This is a class, emit the proper attribute lookup
     if (auto method = classType->findMethod(field)) {
+      if (auto overloaded_methods = classType->findOverloadedMethod(field)) {
```
Right now we return any function that matches the string name of the function unless the name is mangled (then we can return the actual function pointer). Maybe we should error out when this method receives overlaoded method string name without mangling?

tugsbayasgalan committed:
Update on "Method overloading support for Torchbind "


**Approach:**

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea.

Differential Revision: [D27701828](https://our.internmc.facebook.com/intern/diff/D27701828)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27701828).

tugsbayasgalan committed:
Update on "Method overloading support for Torchbind "


**Approach:**

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea.

Differential Revision: [D27701828](https://our.internmc.facebook.com/intern/diff/D27701828)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27701828).

tugsbayasgalan committed:
Update on "Method overloading support for Torchbind "


**Approach:**

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea.

Differential Revision: [D27701828](https://our.internmc.facebook.com/intern/diff/D27701828)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27701828).

SplitInfinity commented:
Okay, I'm satisfied with this PR, but there is one thing I would like a second opinion from @gmagogsfm on.

1. I'm uneasy about storing multiple methods with the same name in the `methods_` member of `ClassType`. As I understand, the reasoning behind this approach is that we do not want to treat regular, non-overloaded methods any differently (than they are being treated today) since overloading is not common. In addition, we cannot tell that a method is going to be overloaded when the first overload is registered, so we cannot decide ahead of time whether to mangle a name or not. The `name` attribute of `Function` is also not mutable. I agree with that reasoning, and there are checks in `findMethod` that return `nullptr` if the method is overloaded, so `findMethod` will not return a random implementation for `name` if multiple exist. But still, having multiple methods with the same name in `methods_` seems like a recipe for disaster.

SplitInfinity commented on torch/csrc/jit/python/pybind_utils.cpp:
```diff
@@ -5,6 +5,26 @@
+    const pybind11::kwargs& kwargs) {
+  auto methods = owner.get_overloaded_methods(method_name);
+  if (methods.has_value()) {
+    for (auto method : methods.value()) {
```
```suggestion
    for (auto& method : methods.value()) {
```

SplitInfinity commented:
I see some minor unrelated formatting changes (e.g. header reordering), it might be a good idea to split them into a separate PR or undo them.

gmagogsfm commented on torch/csrc/jit/frontend/sugared_value.cpp:
```diff
@@ -151,6 +151,10 @@ std::shared_ptr<SugaredValue> SimpleValue::attr(
     // This is a class, emit the proper attribute lookup
+    if (auto overloaded_methods = classType->findOverloadedMethod(field)) {
+      return std::make_shared<MethodValue>(
+          getValue(), overloaded_methods.value());
```
Nit:  constructor of `MethodValue` takes argument by value, not by reference. Consider using `std::move` on `overloaded_methods.value()` to save one copy.

gmagogsfm commented on torch/csrc/jit/python/python_custom_class.cpp:
```diff
@@ -14,15 +16,26 @@ py::object ScriptClass::__call__(py::args args, py::kwargs kwargs) {
+      auto resolved_init_method = match_overloaded_methods(
+          instance._ivalue(), "__init__", input_args, input_kwargs);
+      invokeScriptMethodFromPython(
+          resolved_init_method.value(), input_args, input_kwargs);
```
I think you still need to `std::move` on `input_args` and `input_kwargs` for maximum performance even though you moved them once.

Same for the call on line 38.

gmagogsfm commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -1070,26 +1070,35 @@ struct PythonPrintImpl {
-              method.qualname() ==
-              QualifiedName(selfClass->name()->qualifiedName(), methodName));
+          // check if the method is overloaded and if it is, we serialize
+          // only the qualified name. Since this call happens after first
```
Could you clarify on "this call happens after first round of schema matching"?

tugsbayasgalan replied:
The method name will be mangled because it happens before serialization so this method would only receive mangled name

gmagogsfm replied:
I see, makes sense.

gmagogsfm commented:
> Okay, I'm satisfied with this PR, but there is one thing I would like a second opinion from @gmagogsfm on.
> 
> 1. I'm uneasy about storing multiple methods with the same name in the `methods_` member of `ClassType`. As I understand, the reasoning behind this approach is that we do not want to treat regular, non-overloaded methods any differently (than they are being treated today) since overloading is not common. In addition, we cannot tell that a method is going to be overloaded when the first overload is registered, so we cannot decide ahead of time whether to mangle a name or not. The `name` attribute of `Function` is also not mutable. I agree with that reasoning, and there are checks in `findMethod` that return `nullptr` if the method is overloaded, so `findMethod` will not return a random implementation for `name` if multiple exist. But still, having multiple methods with the same name in `methods_` seems like a recipe for disaster.

Agreed, this is a change of assumption that's made in quite a few places (basically anywhere we iterate through `methods_`). Can we save those overloaded methods in a separate field so existing assumptions are not broken? This separate field would only contain overloaded methods functions pointers, so it should not interfere with existing implementation too much.

tugsbayasgalan commented on aten/src/ATen/core/overloaded_function.h:
```diff
@@ -0,0 +1,184 @@
+
+  std::function<void(Stack&)> callable_;
+
+  c10::FunctionSchema schema_;
```
I deleted this file, i think you might be looking at an older version?

tugsbayasgalan commented on aten/src/ATen/core/custom_class.cpp:
```diff
@@ -34,21 +34,41 @@ bool isCustomClass(const c10::IValue& v) {
 
 void registerCustomClassMethod(std::unique_ptr<jit::Function> fn) {
-  customClassMethods().emplace_back(std::move(fn));
+  auto& custom_class_methods = customClassMethods();
```
Yeahh that is a good point. I will update it

tugsbayasgalan commented on aten/src/ATen/core/custom_class.cpp:
```diff
@@ -34,21 +34,41 @@ bool isCustomClass(const c10::IValue& v) {
+  }
+
+  auto it =
+      custom_class_methods.insert(std::pair<std::string, MethodOverloadsList>(
```
I will delete this code because we don't need to have map anymore

tugsbayasgalan commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -2197,6 +2212,13 @@ struct TORCH_API ClassType : public NamedType {
   std::vector<torch::jit::Function*> staticmethods_;
 
+  // Map of overloaded methods associated with this class that are keyed by
+  // method name
```
yep

tugsbayasgalan commented on aten/src/ATen/core/type.cpp:
```diff
@@ -1086,6 +1087,28 @@ void ClassType::addMethod(torch::jit::Function* method) {
+void ClassType::addOverloadedMethod(torch::jit::Function* method) {
+  // we can't use old findMethod because it searches based on the string name of
+  // function
+  for (auto added_method : methods_) {
```
will do

tugsbayasgalan commented on aten/src/ATen/core/type.cpp:
```diff
@@ -1408,6 +1448,23 @@ torch::jit::Function& ClassType::getMethod(const std::string& name) const {
+
+torch::jit::Function* ClassType::getMangledOverloadedMethod(
+    const std::string& name) const {
+  if (mangled_to_function_.find(name) != mangled_to_function_.end()) {
```
sounds good

tugsbayasgalan commented on aten/src/ATen/core/type.cpp:
```diff
@@ -1408,6 +1448,23 @@ torch::jit::Function& ClassType::getMethod(const std::string& name) const {
 
+c10::optional<std::vector<std::string>> ClassType::findOverloadedMethod(
+    const std::string& name) const {
+  if (overloaded_methods_.find(name) != overloaded_methods_.end()) {
```
sounds good

tugsbayasgalan committed:
Update on "Method overloading support for Torchbind "


**Approach:**

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea.

Differential Revision: [D27701828](https://our.internmc.facebook.com/intern/diff/D27701828)

[ghstack-poisoned]

tugsbayasgalan commented on aten/src/ATen/core/type.cpp:
```diff
@@ -1389,11 +1412,28 @@ void ClassType::checkForwardHookSchema(
+  // only for normal methods, we just return nullptr
+  if (auto overloaded_methods = findOverloadedMethod(name)) {
+    if (overloaded_methods.value().size() == 1) {
+      return getMangledOverloadedMethod(overloaded_methods.value()[0]);
```
There will be **at least** one overload for everything registered through custom class. This is because we don't know if the method will be overloaded or not so we just assume everything is overloaded. (custom_class.h)

gmagogsfm replied:
I see. Could we have a TODO to optimize this? Something like
1) assume method is not overloaded
2) move it to overloaded list when another method with same name registers

tugsbayasgalan replied:
done

tugsbayasgalan commented on aten/src/ATen/core/type.cpp:
```diff
@@ -1086,6 +1087,28 @@ void ClassType::addMethod(torch::jit::Function* method) {
+      method->name() + "__" + c10::guts::to_string(it.first->second.size());
+  it.first->second.push_back(mangled_name);
+  // registers where this overloaded function is stored in the methods map.
+  mangled_to_function_[mangled_name] = methods_.size();
```
sure

tugsbayasgalan replied:
We don't need this for addMethod because it internally calls findMethod which makes sure if there is any overloaded/mangled/normal method with the given name.

tugsbayasgalan committed:
Update on "Method overloading support for Torchbind "


**Approach:**

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea.

Differential Revision: [D27701828](https://our.internmc.facebook.com/intern/diff/D27701828)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27701828).

tugsbayasgalan committed:
Update on "Method overloading support for Torchbind "


**Approach:**

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea.

Differential Revision: [D27701828](https://our.internmc.facebook.com/intern/diff/D27701828)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27701828).

tugsbayasgalan committed:
Update on "Method overloading support for Torchbind "


**Approach:**

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea.

Differential Revision: [D27701828](https://our.internmc.facebook.com/intern/diff/D27701828)

[ghstack-poisoned]

SplitInfinity reviewed (commented):
I have no more comments or suggestions, but am still discomfited by the overall approach (and I don't have any better ideas). Since it has been awhile since we started looking into this, can we revisit the original use case for this to see if it's really worth landing this?

If so, I will approve this PR.

tugsbayasgalan committed:
Update on "Method overloading support for Torchbind "


**Approach:**

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea.

Differential Revision: [D27701828](https://our.internmc.facebook.com/intern/diff/D27701828)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27701828).

tugsbayasgalan committed:
Update on "Method overloading support for Torchbind "


**Approach:**

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea.

Differential Revision: [D27701828](https://our.internmc.facebook.com/intern/diff/D27701828)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27701828).

tugsbayasgalan committed:
Update on "Method overloading support for Torchbind "


**Approach:**

**Building schema from python side**
As you can see from script_init.cpp, we currently load all possible methods for the overloaded function and try to create schema for each of them. And we return whatever that matches first. This approach currently relies on error message that is thrown from createSchemaFromStack method which might not be a good idea.

Differential Revision: [D27701828](https://our.internmc.facebook.com/intern/diff/D27701828)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27701828).
